backend/tracim_backend/lib/utils/dict_parsing.py

```diff
@@ -0,0 +1,120 @@
+import re
+import typing
+
+from marshmallow import ValidationError
+from marshmallow.validate import Regexp
+
+
+def validate_simple_dict(dict_: typing.Dict) -> None:
+    """
+    A simple dict validator with:
+    key as 0-9a-zA-Z-_. based string
+    value as either: string, int, bool, float types
+    """
+    for key in dict_.keys():
+        if not isinstance(key, str):
+            raise ValidationError('Dictionary key "{}" is not a string'.format(key))
+        regex_validator = Regexp(regex=(re.compile("^[0-9a-zA-Z-_.]+$")))
+        try:
+            regex_validator(key)
+        except ValidationError as exc:
+            raise ValidationError(
+                'Dictionary key "{}" incorrect : {}'.format(key, str(exc))
+            ) from exc
+
+    # INFO - G.M - We assume float is the type used for float conversion,
+    # this may change depending
+    # on how the json parser is configured.
+    float_type = float
+    invalid_key_value_pairs = [
+        (key, value)
+        for key, value in dict_.items()
+        if not isinstance(value, (str, int, bool, float_type, type(None)))
+    ]
+    if invalid_key_value_pairs:
+        raise ValidationError(
+            "Only string/number/null values are allowed as dictionary value. Invalid values: {}".format(
+                invalid_key_value_pairs
+            )
+        )
+
+
+def translate_dict(
+    data: dict, keys_to_check: typing.List[str], translation_method: typing.Callable
+) -> dict:
+    """
+    Regenerate similar dict but
+    replace "key: string_value" pair
+    by "key: translation_method(value)" pair
+    recursively
+    :param data: original dict
+    :param keys_to_check: keys value to check for applying translation method
+    :param translation_method: method to run for translation of key
+    :return:
+    """
+    result = {}
+    for key, value in data.items():
+        if isinstance(value, str) and key in keys_to_check:
+            result[key] = translation_method(value)
+        elif isinstance(value, dict):
+            result[key] = translate_dict(
+                data=value, keys_to_check=keys_to_check, translation_method=translation_method,
+            )
+        elif isinstance(value, list):
+            result[key] = translate_list(
+                data=value,
+                keys_to_check=keys_to_check,
+                translation_method=translation_method,
+                key=key,
+            )
+        else:
+            result[key] = value
+    return result
+
+
+def translate_list(
+    data: list,
+    keys_to_check: typing.List[str],
+    translation_method: typing.Callable,
+    key: str = None,
```

Comment:
The test with `key` should be in `translate_dict` and condition the call of this function.
Then `key` can be removed from here.

Reply:
That doesn't seems possible, test is done for every item of the list.

Reply:
- we can replace key there to a "boolean" which say "upper level key is a checked key" but i don't think that necessary.

Reply:
see optimization suggestion few line later.

Reply:
Err… `key` and `keys_to_check` are both given as parameters. Their values do not change in this function, so the condition `key in keys_to_check` is either always `True` or always `False` whatever the list item.
Or do I miss something?